program-test/src/programs.rs

```diff
@@ -39,48 +40,91 @@ static SPL_PROGRAMS: &[(Pubkey, Pubkey, &[u8])] = &[
     ),
 ];
 
+/// Returns a tuple `(Pubkey, Account)` for a BPF program, where the key is the
+/// provided program ID and the account is a valid BPF Loader program account
+/// containing the ELF.
+pub fn bpf_loader_program_account(
+    program_id: &Pubkey,
+    elf: &[u8],
+    rent: &Rent,
+) -> (Pubkey, Account) {
+    (
+        *program_id,
+        Account {
+            lamports: rent.minimum_balance(elf.len()).max(1),
+            data: elf.to_vec(),
+            owner: bpf_loader::id(),
+            executable: true,
+            rent_epoch: 0,
+        },
+    )
+}
+
+/// Returns two tuples `(Pubkey, Account)` for a BPF upgradeable program.
+/// The first tuple is the program account. It contains the provided program ID
+/// and an account with a pointer to its program data account.
+/// The second tuple is the program data account. It contains the program data
+/// address and an account with the program data - a valid BPF Loader Upgradeable
+/// program data account containing the ELF.
+pub fn bpf_loader_upgradeable_program_accounts(
```

Comment:
Unless there's a use-case I'm missing, it might be a bit confusing for external callers to use this function -- how about making this `pub(crate)` and adding a helper function on `ProgramTest` which will add a program given an elf file?

It could be `ProgramTest::add_upgradeable_program(program_id: &Pubkey, elf: &[u8])` and use this inside -- what do you think?

Reply:
I went this route to add the minimal "niceness" to the API and offer more generic tooling downstream, but you're right maybe we don't have to.

For our use cases we'd need this bad boy to look something like `ProgramTest::add_upgradeable_program_to_genesis(/* ... */)`.

If we're going to add it to `ProgramTest` directly, what do you think about getting the fancy ELF file finding out-of-the-box?

```
fn add_upgradeable_program_to_genesis(program_id: &Pubkey, program_name: &str);
```

This way our usage becomes:

```rust
let mut program_test = ProgramTest::default();
program_test.add_upgradeable_program_to_genesis(
    "solana_address_lookup_table_program",
    solana_address_lookup_table_program::id(),
);
program_test.start_with_context().await
```

Reply:
I've added a commit for just wrapping the file reading, but we can pop it if we don't want it.

Reply:
Yeah that works! Since we already have `add_program`, how about just `add_upgradeable_program` to keep it simple?

Reply:
I think we should be making a distinction that it's adding it to genesis, unlike `add_program` or `add_account`, which will add to an existing bank.

In 99% of cases it won't matter I guess...

Reply:
Ah right. For my own information, why does it need to be added to genesis?

Reply:
Basically we need to get in front of this bad boy.

https://github.com/anza-xyz/agave/blob/c038908600b8a1b0080229dea015d7fc9939c418/runtime/src/bank.rs#L5109-L5126

On `start`, Program-Test will initialize a Bank and then add the accounts, but if you go through genesis then Bank's `finish_init` has the program accounts already when it's initializing itself.